lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp

```diff
@@ -140,6 +140,54 @@ class StableHLOToTTIRTransposeOpConversionPattern
         srcOp, outputTensor.getType(), Value(adaptor.getOperand()),
         Value(outputTensor), adaptor.getPermutation()[0],
         adaptor.getPermutation()[1],
+
+        rewriter.getArrayAttr(
+            SmallVector<Attribute>(adaptor.getOperands().size() + 1,
+                                   rewriter.getAttr<OperandConstraintAttr>(
+                                       OperandConstraint::AnyDeviceTile))));
+    return success();
+  }
+};
+
+class StableHLOToTTIRDotGeneralOpConversionPattern
+    : public OpConversionPattern<mlir::stablehlo::DotGeneralOp> {
+  using OpConversionPattern<mlir::stablehlo::DotGeneralOp>::OpConversionPattern;
+
+public:
+  LogicalResult
+  matchAndRewrite(mlir::stablehlo::DotGeneralOp srcOp,
+                  mlir::stablehlo::DotGeneralOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto outputType = mlir::cast<RankedTensorType>(srcOp.getResult().getType());
+    auto outputTensor = rewriter.create<tensor::EmptyOp>(
+        srcOp.getLoc(), outputType.getShape(), outputType.getElementType());
+
+    // This is a basic version that can only work for cases that can be directly
+    // converted to matmul. The op should be extended as other ops such as
+    // ttir.permute and ttir.broadcast_in_dim become available.
+
+    ::mlir::stablehlo::DotDimensionNumbersAttr dimensions =
+        adaptor.getDotDimensionNumbers();
+    if (dimensions.getLhsContractingDimensions().size()) {
```

Comment:
You should also return failure here when this is empty. Change it to something like this (when you move this to `checkBasicLegality` function):
```
if (dimensions.getLhsContractingDimensions().size() != 1 ||
    dimensions.getLhsContractingDimensions()[0] != 1) {
  // Only regular non-transposed matmul is supported currently.
  return false;
}
```

Similar for the next `if` below.

Reply:
I will move the error checking to a separate function like `checkBasicLegality`

Reply:
This wasn't fixed yet when you moved the code to `checkBasicLegality`

Reply:
In any case it is a good practice to check for the size of the external container before indexing it